source/geod24/concurrency.d

```diff
@@ -1691,19 +1828,32 @@ public class Channel (T)
         assert(msg !is null);
 
         this.mutex.lock();
-        scope (exit)
-            this.mutex.unlock();
 
         if (this.closed)
+        {
+            this.mutex.unlock();
             return false;
+        }
+
+        if (!this.sendq.empty)
+        {
+            ChannelContext!T context = this.sendq.front;
+            this.sendq.removeFront();
+            *(msg) = context.msg;
+            this.mutex.unlock();
+            context.notify();
+            return true;
+        }
 
         if (!this.queue.empty)
         {
             *(msg) = this.queue.front;
             this.queue.removeFront();
```

Comment:
There is a missing `this.mcount--;` decrement here. The tests need to be improved because they didn't catch this error.

```suggestion
            this.queue.removeFront();
            assert(this.mcount > 0, "mcount was zero!");
            this.mcount--;
```